local TypeUtils = require("../Source/Commandline/Helpers/TypeUtils")

return function()
    describe("TypeUtils.default", function()
        it("should return value when it matches defaultValue type", function()
            expect(TypeUtils.default(42, 0)).to.equal(42)
            expect(TypeUtils.default("hello", "")).to.equal("hello")
            expect(TypeUtils.default(true, false)).to.equal(true)
        end)

        it("should return defaultValue when value is nil", function()
            expect(TypeUtils.default(nil, 42)).to.equal(42)
            expect(TypeUtils.default(nil, "default")).to.equal("default")
            expect(TypeUtils.default(nil, false)).to.equal(false)
        end)

        it("should return defaultValue when value type does not match", function()
            expect(TypeUtils.default("string", 42)).to.equal(42)
            expect(TypeUtils.default(123, "default")).to.equal("default")
            expect(TypeUtils.default("true", false)).to.equal(false)
        end)
    end)

    describe("TypeUtils.isValidType", function()
        it("should return true for valid types", function()
            expect(TypeUtils.isValidType("boolean")).to.equal(true)
            expect(TypeUtils.isValidType("int")).to.equal(true)
            expect(TypeUtils.isValidType("number")).to.equal(true)
            expect(TypeUtils.isValidType("string")).to.equal(true)
        end)

        it("should return false for invalid types", function()
            expect(TypeUtils.isValidType("invalid")).to.equal(false)
            expect(TypeUtils.isValidType("table")).to.equal(false)
            expect(TypeUtils.isValidType("function")).to.equal(false)
            expect(TypeUtils.isValidType("Boolean")).to.equal(false)
            expect(TypeUtils.isValidType("")).to.equal(false)
        end)
    end)

    describe("TypeUtils.validateValueType", function()
        describe("boolean type", function()
            it("should validate boolean values correctly", function()
                expect(TypeUtils.validateValueType(true, "boolean")).to.equal(true)
                expect(TypeUtils.validateValueType(false, "boolean")).to.equal(true)
            end)

            it("should reject non-boolean values", function()
                expect(TypeUtils.validateValueType("true", "boolean")).to.equal(false)
                expect(TypeUtils.validateValueType(1, "boolean")).to.equal(false)
                expect(TypeUtils.validateValueType(nil, "boolean")).to.equal(false)
            end)
        end)

        describe("int type", function()
            it("should validate integer values correctly", function()
                expect(TypeUtils.validateValueType(0, "int")).to.equal(true)
                expect(TypeUtils.validateValueType(42, "int")).to.equal(true)
                expect(TypeUtils.validateValueType(-10, "int")).to.equal(true)
            end)

            it("should reject non-integer values", function()
                expect(TypeUtils.validateValueType(3.14, "int")).to.equal(false)
                expect(TypeUtils.validateValueType(-2.5, "int")).to.equal(false)
                expect(TypeUtils.validateValueType("42", "int")).to.equal(false)
            end)
        end)

        describe("number type", function()
            it("should validate number values correctly", function()
                expect(TypeUtils.validateValueType(0, "number")).to.equal(true)
                expect(TypeUtils.validateValueType(42, "number")).to.equal(true)
                expect(TypeUtils.validateValueType(3.14, "number")).to.equal(true)
                expect(TypeUtils.validateValueType(-2.5, "number")).to.equal(true)
            end)

            it("should reject non-number values", function()
                expect(TypeUtils.validateValueType("42", "number")).to.equal(false)
                expect(TypeUtils.validateValueType(true, "number")).to.equal(false)
            end)
        end)

        describe("string type", function()
            it("should validate string values correctly", function()
                expect(TypeUtils.validateValueType("", "string")).to.equal(true)
                expect(TypeUtils.validateValueType("hello", "string")).to.equal(true)
                expect(TypeUtils.validateValueType("123", "string")).to.equal(true)
            end)

            it("should reject non-string values", function()
                expect(TypeUtils.validateValueType(42, "string")).to.equal(false)
                expect(TypeUtils.validateValueType(true, "string")).to.equal(false)
            end)
        end)
    end)

    describe("TypeUtils.castFromString", function()
        describe("boolean type", function()
            it("should cast 'true' to true", function()
                local value, err = TypeUtils.castFromString("true", "boolean")
                expect(err).to.equal(nil)
                expect(value).to.equal(true)
            end)

            it("should cast 'false' to false", function()
                local value, err = TypeUtils.castFromString("false", "boolean")
                expect(err).to.equal(nil)
                expect(value).to.equal(false)
            end)

            it("should be case insensitive", function()
                local value1, err1 = TypeUtils.castFromString("TRUE", "boolean")
                expect(err1).to.equal(nil)
                expect(value1).to.equal(true)

                local value2, err2 = TypeUtils.castFromString("False", "boolean")
                expect(err2).to.equal(nil)
                expect(value2).to.equal(false)
            end)

            it("should return error for invalid boolean strings", function()
                local value, err = TypeUtils.castFromString("yes", "boolean")
                expect(value).to.equal(nil)
                expect(err).to.be.ok()
                expect(err:find("Unable to cast")).to.be.ok()
            end)
        end)

        describe("int type", function()
            it("should cast valid integer strings", function()
                local value1, err1 = TypeUtils.castFromString("42", "int")
                expect(err1).to.equal(nil)
                expect(value1).to.equal(42)

                local value2, err2 = TypeUtils.castFromString("-10", "int")
                expect(err2).to.equal(nil)
                expect(value2).to.equal(-10)

                local value3, err3 = TypeUtils.castFromString("0", "int")
                expect(err3).to.equal(nil)
                expect(value3).to.equal(0)
            end)

            it("should return error for non-numeric strings", function()
                local value, err = TypeUtils.castFromString("abc", "int")
                expect(value).to.equal(nil)
                expect(err).to.be.ok()
                expect(err:find("Not a valid number")).to.be.ok()
            end)

            it("should return error for non-integer numbers", function()
                local value, err = TypeUtils.castFromString("3.14", "int")
                expect(value).to.equal(nil)
                expect(err).to.be.ok()
                expect(err:find("Expected an integer")).to.be.ok()
            end)
        end)

        describe("number type", function()
            it("should cast valid number strings", function()
                local value1, err1 = TypeUtils.castFromString("42", "number")
                expect(err1).to.equal(nil)
                expect(value1).to.equal(42)

                local value2, err2 = TypeUtils.castFromString("3.14", "number")
                expect(err2).to.equal(nil)
                expect(value2).to.equal(3.14)

                local value3, err3 = TypeUtils.castFromString("-2.5", "number")
                expect(err3).to.equal(nil)
                expect(value3).to.equal(-2.5)
            end)

            it("should return error for non-numeric strings", function()
                local value, err = TypeUtils.castFromString("not-a-number", "number")
                expect(value).to.equal(nil)
                expect(err).to.be.ok()
                expect(err:find("Not a valid number")).to.be.ok()
            end)
        end)

        describe("string type", function()
            it("should return the string as-is", function()
                local value1, err1 = TypeUtils.castFromString("hello", "string")
                expect(err1).to.equal(nil)
                expect(value1).to.equal("hello")

                local value2, err2 = TypeUtils.castFromString("123", "string")
                expect(err2).to.equal(nil)
                expect(value2).to.equal("123")

                local value3, err3 = TypeUtils.castFromString("", "string")
                expect(err3).to.equal(nil)
                expect(value3).to.equal("")
            end)
        end)

        describe("unknown type", function()
            it("should return error for unknown types", function()
                local value, err = TypeUtils.castFromString("test", "unknown")
                expect(value).to.equal(nil)
                expect(err).to.be.ok()
                expect(err:find("Unknown type")).to.be.ok()
            end)
        end)
    end)
end
